#'---
#' title: RNA Variant Calling
#' author:
#' wb:
#'  log:
#'    - snakemake: '`sm str(tmp_dir / "RVC" / "Overview.Rds")`'
#'  params:
#'    - rna_ids: '`sm cfg.RVC.batchIDs`'
#'    - ids: '`sm cfg.sampleAnnotation.subsetSampleAnnotation("RNA_ID",cfg.RVC.batchIDs)[["RNA_ID","DNA_ID","DNA_VCF_FILE"]].fillna("").values.tolist()`'
#'    - dna_vcfs: '`sm cfg.sampleAnnotation.getFilePath(cfg.RVC.batchIDs,"DNA_VCF_FILE",single_file = False)`'
#'  input:
#'    - helper_methods: '`sm cfg.workDir / "Scripts/rnaVariantCalling/pipeline/RVC_helpers/RVC_helper.R"`'
#'    - vcfFilesBatch: '`sm expand(cfg.getProcessedDataDir() +
#'                      "/rnaVariantCalling/out/all_samples_haplocaller/" + 
#'                      "{batchID}_all_samples.genotyped.filtered_clean.vcf.gz",
#'                  batchID=cfg.RVC.groups)`'
#'    - vcfFiles: '`sm expand(cfg.getProcessedDataDir() +
#'                      "/rnaVariantCalling/out/sample_haplocaller/" + 
#'                      "{sample}/{sample}.genotyped.filtered.basic{min_alt}.vcf.gz",
#'                  sample=cfg.RVC.batchIDs,min_alt=getMinAlt())`'
#'    - vcfFilesMasked: '`sm expand(cfg.getProcessedDataDir() +
#'                      "/rnaVariantCalling/out/sample_haplocaller/" + 
#'                      "{sample}/{sample}.genotyped.filtered.basic{min_alt}.masked.vcf.gz",
#'                  sample=cfg.RVC.batchIDs,min_alt=getMinAlt())`'
#' output:
#'   html_document:
#'    code_folding: hide
#'    code_download: TRUE
#'---

library(data.table)
library(tidyr)
library(dplyr)
library(vcfR)
saveRDS(snakemake, snakemake@log$snakemake)
#snakemake <- readRDS("/home/nicksmith/Documents/projects/drop_work/demo3/.drop/tmp/RVC/Overview.Rds")
source(snakemake@input$helper_methods)

#subset sample_annotation.tsv and match RVC IDs with DNA information (if available)
sample_table <- parse_rna_dna_names(snakemake@params$ids)

# merge with the RNA_VCFs generated by their RNA_ID
vcf_sample_table <- data.table("RNA_VCF_FILE" = snakemake@input$vcfFiles,"RNA_ID" = names(snakemake@params$rna_ids))
vcf_sample_table <- merge.data.table(vcf_sample_table,sample_table,by = "RNA_ID",all.x = T)

all_samples <- lapply(1:nrow(vcf_sample_table),function(i) {
       extract_vcf_to_file (
           rnaseq_vcf_file = as.character(vcf_sample_table[i,"RNA_VCF_FILE"]),
           exome_vcf_file  = as.character(vcf_sample_table[i,"DNA_VCF_FILE"]),
           rna_name        = as.character(vcf_sample_table[i,"RNA_ID"]),
           wes_name        = as.character(vcf_sample_table[i,"DNA_ID"]))
       })

names(all_samples) <- names(snakemake@params$rna_ids)
all_variants <- rbindlist(all_samples,fill = T,idcol = T)
all_variants <- dplyr::rename(all_variants,"SAMPLE" = ".id")

