{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno = pd.read_csv(\"../../../sample_annotation/Data/sample_annotation.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'100416'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print(anno['FIBROBLAST_ID'])\n",
    "# anno.iloc[1]\n",
    "anno.loc[1,\"FIBROBLAST_ID\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_mae = anno[anno[\"LAB\"] == \"PROKISCH\"] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1268, 53)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(anno_mae[\"LAB\"])\n",
    "anno_mae.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#np.isnan(anno_mae.EXOME_ID)\n",
    "anno_mae = anno_mae[pd.notnull(anno_mae.EXOME_ID)]\n",
    "anno_mae = anno_mae[pd.notnull(anno_mae.RNA_ID)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(284, 53)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno_mae.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_mae = anno_mae[[\"EXOME_ID\", \"RNA_ID\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_mae['rna_file'] = [\"/s/project/mitoMultiOmics/raw_data/helmholtz/\" + x + \"/RNAout/paired-endout/stdFilenames/\" + x + \".bam\" for x in anno_mae[\"RNA_ID\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_mae['vcf_file'] = [\"/s/project/mitoMultiOmics/raw_data/helmholtz/\" + x + \"/exomicout/paired-endout/stdFilenames/\" + x + \".vcf.gz\" for x in anno_mae[\"EXOME_ID\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# anno_mae[\"vcf_file\"]\n",
    "anno_mae['vcf_exists'] = [os.path.exists(x) for x in anno_mae[\"vcf_file\"]]\n",
    "anno_mae['rna_exists'] = [os.path.exists(x) for x in anno_mae[\"rna_file\"]]\n",
    "anno_mae = anno_mae[anno_mae['vcf_exists'] & anno_mae['rna_exists']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ('mae/' + anno_mae[\"EXOME_ID\"] + '-' + anno_mae[\"RNA_ID\"] + '.Rds').tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Complete Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mae_files(anno):\n",
    "    \n",
    "    # subset and clean\n",
    "    anno_mae = anno[anno[\"LAB\"] == \"PROKISCH\"]\n",
    "    anno_mae = anno_mae[pd.notnull(anno_mae.EXOME_ID)]\n",
    "    anno_mae = anno_mae[pd.notnull(anno_mae.RNA_ID)]\n",
    "    anno_mae = anno_mae[[\"EXOME_ID\", \"RNA_ID\"]]\n",
    "\n",
    "    # create file names\n",
    "    anno_mae['rna_file'] = [\"/s/project/mitoMultiOmics/raw_data/helmholtz/\" + x + \"/RNAout/paired-endout/stdFilenames/\" + x + \".bam\" for x in anno_mae[\"RNA_ID\"]]\n",
    "    anno_mae['vcf_file'] = [\"/s/project/mitoMultiOmics/raw_data/helmholtz/\" + x + \"/exomicout/paired-endout/stdFilenames/\" + x + \".vcf.gz\" for x in anno_mae[\"EXOME_ID\"]]\n",
    "\n",
    "    # check for missing files\n",
    "    anno_mae['vcf_exists'] = [os.path.exists(x) for x in anno_mae[\"vcf_file\"]]\n",
    "    anno_mae['rna_exists'] = [os.path.exists(x) for x in anno_mae[\"rna_file\"]]\n",
    "    anno_mae = anno_mae[anno_mae['vcf_exists'] & anno_mae['rna_exists']]\n",
    "    \n",
    "    vcf = anno_mae[\"EXOME_ID\"] \n",
    "    rna = anno_mae[\"RNA_ID\"]\n",
    "    \n",
    "    return vcf.tolist(), rna.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf, rna = mae_files(anno)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test for all vcf files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_vcf(anno):\n",
    "\n",
    "    # subset and clean\n",
    "    anno = anno[anno[\"LAB\"] == \"PROKISCH\"]\n",
    "    anno = anno[pd.notnull(anno.EXOME_ID)]\n",
    "    anno_vcf = anno[[\"EXOME_ID\"]].copy()\n",
    "\n",
    "    anno_vcf['file'] = [\"/s/project/mitoMultiOmics/raw_data/helmholtz/\" + x + \"/exomicout/paired-endout/stdFilenames/\" + x + \".vcf.gz\" for x in anno_vcf[\"EXOME_ID\"]]\n",
    "    anno_vcf['vcf_exists'] = [os.path.exists(x) for x in anno_vcf[\"file\"]]\n",
    "    anno_vcf = anno_vcf[anno_vcf['vcf_exists']]\n",
    "    \n",
    "    return pd.unique(anno_vcf[\"EXOME_ID\"]).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "752"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_vcfs = all_vcf(anno)\n",
    "len(all_vcfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "232"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(vcf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## OUTRIDER group files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno = pd.read_csv(\"../../../sample_annotation/Data/sample_annotation.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FIBROBLAST_ID</th>\n",
       "      <th>EXOME_ID</th>\n",
       "      <th>RNA_ID</th>\n",
       "      <th>GENOME_ID</th>\n",
       "      <th>KNOWN_MUTATION</th>\n",
       "      <th>TISSUE</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>MAX_RNA_TO_EXOME_IDENTITY</th>\n",
       "      <th>PROTEOME_ID</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>...</th>\n",
       "      <th>IS_RNA_SEQ_STRANDED</th>\n",
       "      <th>BATCH</th>\n",
       "      <th>PubMed_ID</th>\n",
       "      <th>Last_revised</th>\n",
       "      <th>fib_id</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>PEDIGREE_new</th>\n",
       "      <th>OUTRIDER_GROUP</th>\n",
       "      <th>CANDIDATE</th>\n",
       "      <th>remove</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MUC3196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "      <td>HEALTHY</td>\n",
       "      <td>0.41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>all,fib_ss,fib,no_blood</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MUC3197</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "      <td>HEALTHY</td>\n",
       "      <td>0.41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>all,gal,fib,no_blood</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100634</td>\n",
       "      <td>92582</td>\n",
       "      <td>100634R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "      <td>MITO</td>\n",
       "      <td>0.94</td>\n",
       "      <td>P100634</td>\n",
       "      <td>female</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20181119.0</td>\n",
       "      <td>100634</td>\n",
       "      <td>p67077</td>\n",
       "      <td>ExDiag0510</td>\n",
       "      <td>all,fib_ss,fib,no_blood</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100782</td>\n",
       "      <td>98288</td>\n",
       "      <td>110468R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "      <td>MITO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P100782</td>\n",
       "      <td>male</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Batch_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100782</td>\n",
       "      <td>p70754</td>\n",
       "      <td>ExDiag0874</td>\n",
       "      <td>all,fib_ss,fib,no_blood</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>101309</td>\n",
       "      <td>104905</td>\n",
       "      <td>110442R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "      <td>MITO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Batch_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>101309</td>\n",
       "      <td>p72867</td>\n",
       "      <td>SALK0009</td>\n",
       "      <td>all,fib_ss,fib,no_blood</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  FIBROBLAST_ID EXOME_ID   RNA_ID GENOME_ID KNOWN_MUTATION      TISSUE  \\\n",
       "0        100416      NaN  MUC3196       NaN            NaN  FIBROBLAST   \n",
       "1        100416      NaN  MUC3197       NaN            NaN  FIBROBLAST   \n",
       "2        100634    92582  100634R       NaN            NaN  FIBROBLAST   \n",
       "3        100782    98288  110468R       NaN            NaN  FIBROBLAST   \n",
       "4        101309   104905  110442R       NaN            NaN  FIBROBLAST   \n",
       "\n",
       "   DISEASE  MAX_RNA_TO_EXOME_IDENTITY PROTEOME_ID  GENDER  ...  \\\n",
       "0  HEALTHY                       0.41         NaN     NaN  ...   \n",
       "1  HEALTHY                       0.41         NaN     NaN  ...   \n",
       "2     MITO                       0.94     P100634  female  ...   \n",
       "3     MITO                        NaN     P100782    male  ...   \n",
       "4     MITO                        NaN         NaN  female  ...   \n",
       "\n",
       "  IS_RNA_SEQ_STRANDED    BATCH PubMed_ID Last_revised  fib_id PATIENT_ID  \\\n",
       "0                 1.0  Batch_2       NaN          NaN  100416        NaN   \n",
       "1                 1.0  Batch_2       NaN          NaN  100416        NaN   \n",
       "2                 1.0  Batch_2       NaN   20181119.0  100634     p67077   \n",
       "3                 1.0  Batch_4       NaN          NaN  100782     p70754   \n",
       "4                 1.0  Batch_4       NaN          NaN  101309     p72867   \n",
       "\n",
       "   PEDIGREE_new           OUTRIDER_GROUP  CANDIDATE  remove  \n",
       "0           NaN  all,fib_ss,fib,no_blood        NaN     NaN  \n",
       "1           NaN     all,gal,fib,no_blood        NaN     NaN  \n",
       "2    ExDiag0510  all,fib_ss,fib,no_blood        NaN     NaN  \n",
       "3    ExDiag0874  all,fib_ss,fib,no_blood        NaN     NaN  \n",
       "4      SALK0009  all,fib_ss,fib,no_blood        NaN     NaN  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outrider_files(anno):\n",
    "\n",
    "    # subset and clean\n",
    "    anno_outrider = anno[(anno.LAB == \"PROKISCH\") & pd.notnull(anno.RNA_ID) & pd.notnull(anno.OUTRIDER_GROUP)]\n",
    "    anno_outrider = anno_outrider[[\"RNA_ID\", \"OUTRIDER_GROUP\"]].drop_duplicates().copy()\n",
    "\n",
    "    # create filenames and ignore missing files\n",
    "    anno_outrider['file'] = [\"/s/project/mitoMultiOmics/raw_data/helmholtz/\" + x + \"/RNAout/\"\n",
    "                             for x in anno_outrider[\"RNA_ID\"]]\n",
    "    anno_outrider['file_exists'] = [os.path.exists(x) for x in anno_outrider[\"file\"]]\n",
    "    anno_outrider = anno_outrider[anno_outrider['file_exists']]\n",
    "\n",
    "    # subset by OUTRIDER_GROUP\n",
    "    outrider_groups = []\n",
    "    for s in set(anno_outrider.OUTRIDER_GROUP):\n",
    "        outrider_groups.extend(s.split(','))\n",
    "    outrider_ids = {og : anno_outrider.loc[anno_outrider.OUTRIDER_GROUP.str.contains('(^|,)' + og + '(,|$)'), 'RNA_ID'].tolist() for og in set(outrider_groups)}\n",
    "    return {og: _list for og, _list in outrider_ids.items() if len(_list) > 40}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/3-4.1.1/envs/genetic_diagnosis/lib/python3.6/site-packages/ipykernel_launcher.py:17: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.\n"
     ]
    }
   ],
   "source": [
    "of = outrider_files(anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "665"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(of['all'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(345, 345)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(of['fib']), len(of['fib_all'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_outrider = anno[(anno.LAB == \"PROKISCH\") & pd.notnull(anno.RNA_ID) & pd.notnull(anno.OUTRIDER_GROUP)]\n",
    "anno_outrider = anno_outrider[[\"RNA_ID\", \"OUTRIDER_GROUP\"]].drop_duplicates().copy()\n",
    "\n",
    "    # create filenames and ignore missing files\n",
    "anno_outrider['file'] = [\"/s/project/mitoMultiOmics/raw_data/helmholtz/\" + x + \"/RNAout/\"\n",
    "                             for x in anno_outrider[\"RNA_ID\"]]\n",
    "anno_outrider['file_exists'] = [os.path.exists(x) for x in anno_outrider[\"file\"]]\n",
    "anno_outrider = anno_outrider[anno_outrider['file_exists']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA_ID</th>\n",
       "      <th>OUTRIDER_GROUP</th>\n",
       "      <th>file</th>\n",
       "      <th>file_exists</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MUC3196</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MUC3197</td>\n",
       "      <td>all,gal,fib_all,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100634R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110468R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110442R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>110460R</td>\n",
       "      <td>all,blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>101401R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>101676R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>102384R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>102385R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>102386R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>102387R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>102388R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>102389R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>102390R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>102391R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>102392R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>102393R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>102630R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>102631R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>102632R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>102860R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>102864R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>103174R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>102875R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>102876R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>102949R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>102950R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>102951R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>102952R</td>\n",
       "      <td>all,fib_ss,fib_all,fib,no_blood</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/1...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2371</th>\n",
       "      <td>MUC2170</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2372</th>\n",
       "      <td>MUC2171</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373</th>\n",
       "      <td>MUC2173</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2374</th>\n",
       "      <td>MUC2174</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375</th>\n",
       "      <td>MUC2175</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2376</th>\n",
       "      <td>MUC2176</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>MUC2177</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>MUC2178</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>MUC2179</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>MUC2180</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2381</th>\n",
       "      <td>MUC2181</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>MUC2182</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2383</th>\n",
       "      <td>MUC2183</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2384</th>\n",
       "      <td>MUC2184</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2385</th>\n",
       "      <td>MUC2185</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2386</th>\n",
       "      <td>MUC2186</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2387</th>\n",
       "      <td>MUC2187</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388</th>\n",
       "      <td>MUC2188</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389</th>\n",
       "      <td>MUC2189</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2390</th>\n",
       "      <td>MUC2190</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2391</th>\n",
       "      <td>MUC2191</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2392</th>\n",
       "      <td>MUC2192</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>MUC2193</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>MUC2194</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>MUC2195</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>MUC2196</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>MUC2197</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2398</th>\n",
       "      <td>MUC2198</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2399</th>\n",
       "      <td>MUC2199</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2400</th>\n",
       "      <td>MUC2200</td>\n",
       "      <td>all</td>\n",
       "      <td>/s/project/mitoMultiOmics/raw_data/helmholtz/M...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>665 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       RNA_ID                   OUTRIDER_GROUP  \\\n",
       "0     MUC3196  all,fib_ss,fib_all,fib,no_blood   \n",
       "1     MUC3197         all,gal,fib_all,no_blood   \n",
       "2     100634R  all,fib_ss,fib_all,fib,no_blood   \n",
       "3     110468R  all,fib_ss,fib_all,fib,no_blood   \n",
       "4     110442R  all,fib_ss,fib_all,fib,no_blood   \n",
       "5     110460R                        all,blood   \n",
       "6     101401R  all,fib_ss,fib_all,fib,no_blood   \n",
       "7     101676R  all,fib_ss,fib_all,fib,no_blood   \n",
       "8     102384R  all,fib_ss,fib_all,fib,no_blood   \n",
       "9     102385R  all,fib_ss,fib_all,fib,no_blood   \n",
       "10    102386R  all,fib_ss,fib_all,fib,no_blood   \n",
       "11    102387R  all,fib_ss,fib_all,fib,no_blood   \n",
       "12    102388R  all,fib_ss,fib_all,fib,no_blood   \n",
       "13    102389R  all,fib_ss,fib_all,fib,no_blood   \n",
       "14    102390R  all,fib_ss,fib_all,fib,no_blood   \n",
       "15    102391R  all,fib_ss,fib_all,fib,no_blood   \n",
       "16    102392R  all,fib_ss,fib_all,fib,no_blood   \n",
       "17    102393R  all,fib_ss,fib_all,fib,no_blood   \n",
       "18    102630R  all,fib_ss,fib_all,fib,no_blood   \n",
       "19    102631R  all,fib_ss,fib_all,fib,no_blood   \n",
       "20    102632R  all,fib_ss,fib_all,fib,no_blood   \n",
       "21    102860R  all,fib_ss,fib_all,fib,no_blood   \n",
       "22    102864R  all,fib_ss,fib_all,fib,no_blood   \n",
       "23    103174R  all,fib_ss,fib_all,fib,no_blood   \n",
       "24    102875R  all,fib_ss,fib_all,fib,no_blood   \n",
       "25    102876R  all,fib_ss,fib_all,fib,no_blood   \n",
       "26    102949R  all,fib_ss,fib_all,fib,no_blood   \n",
       "27    102950R  all,fib_ss,fib_all,fib,no_blood   \n",
       "28    102951R  all,fib_ss,fib_all,fib,no_blood   \n",
       "30    102952R  all,fib_ss,fib_all,fib,no_blood   \n",
       "...       ...                              ...   \n",
       "2371  MUC2170                              all   \n",
       "2372  MUC2171                              all   \n",
       "2373  MUC2173                              all   \n",
       "2374  MUC2174                              all   \n",
       "2375  MUC2175                              all   \n",
       "2376  MUC2176                              all   \n",
       "2377  MUC2177                              all   \n",
       "2378  MUC2178                              all   \n",
       "2379  MUC2179                              all   \n",
       "2380  MUC2180                              all   \n",
       "2381  MUC2181                              all   \n",
       "2382  MUC2182                              all   \n",
       "2383  MUC2183                              all   \n",
       "2384  MUC2184                              all   \n",
       "2385  MUC2185                              all   \n",
       "2386  MUC2186                              all   \n",
       "2387  MUC2187                              all   \n",
       "2388  MUC2188                              all   \n",
       "2389  MUC2189                              all   \n",
       "2390  MUC2190                              all   \n",
       "2391  MUC2191                              all   \n",
       "2392  MUC2192                              all   \n",
       "2393  MUC2193                              all   \n",
       "2394  MUC2194                              all   \n",
       "2395  MUC2195                              all   \n",
       "2396  MUC2196                              all   \n",
       "2397  MUC2197                              all   \n",
       "2398  MUC2198                              all   \n",
       "2399  MUC2199                              all   \n",
       "2400  MUC2200                              all   \n",
       "\n",
       "                                                   file  file_exists  \n",
       "0     /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "1     /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "3     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "4     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "5     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "6     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "7     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "8     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "9     /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "10    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "11    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "12    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "13    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "14    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "15    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "16    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "17    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "18    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "19    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "20    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "21    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "22    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "23    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "24    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "25    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "26    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "27    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "28    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "30    /s/project/mitoMultiOmics/raw_data/helmholtz/1...         True  \n",
       "...                                                 ...          ...  \n",
       "2371  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2372  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2373  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2374  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2375  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2376  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2377  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2378  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2379  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2380  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2381  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2382  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2383  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2384  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2385  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2386  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2387  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2388  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2389  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2390  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2391  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2392  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2393  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2394  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2395  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2396  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2397  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2398  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2399  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "2400  /s/project/mitoMultiOmics/raw_data/helmholtz/M...         True  \n",
       "\n",
       "[665 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno_outrider"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/3-4.1.1/envs/genetic_diagnosis/lib/python3.6/site-packages/ipykernel_launcher.py:4: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "outrider_groups = []\n",
    "for s in set(anno_outrider.OUTRIDER_GROUP):\n",
    "    outrider_groups.extend(s.split(','))\n",
    "outrider_ids = {og : anno_outrider.loc[anno_outrider.OUTRIDER_GROUP.str.contains('(^|,)' + og + '(,|$)'), 'RNA_ID'].tolist() for og in set(outrider_groups)}\n",
    "#{og: _list for og, _list in outrider_ids.items() if len(_list) > 40}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
